keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,280 @@
++ _Static_ - also called "static brownfield" - buckets are created outside the COSI system but granted access via COSI without the need for the driver/sidecar. The OB and driver secret are created manually in the expected locations. The app secret and binding are performed by COSI.
+
+# Proposal
+
+## User Stories
+
+#### Admin
+
+- As a cluster administrator, I can set quotas and resource limits on generated buckets' storage capacity via the Kubernete's API, so that  I can control monthly infrastructure costs.
+- As a cluster administrator, I can use Kubernetes RBAC policies on bucket APIs, so that I may control integration and access to pre-existing buckets from within the cluster, reducing the need to administer an external storage interface.
+- As a cluster administrator, I can manage multiple object store providers via the Kubernetes API, so that I do not have to become an expert in several different storage interfaces.
+
+#### User
+
+- As a developer, I can define my object storage needs in the same manifest as my workload, so that deployments are streamlined and encapsulated within the Kubernetes interface.
+- As a developer, I can define a manifest containing my workload and object storage configuration once, so that my app may be ported between clusters as long as the storage provided supports my designated data path protocol.
+
+  
+
+## System Configuration
+
++ The central controller runs in a protected namespace with RBAC privileges for managing OBCs, OBs, and Secrets cluster wide.
+    + Responsible for the binding relationship of OBs and OBCs.
++ The "cosi-system" namespace must be created.
+    + This namespace must be granted write access to cluster-wide scoped OBs.
++ The Driver and Sidecar containers run together in a single Pod in the "cosi-system" namespace.
+    + The gRPC connection is made through the Pod’s localhost.
++ BucketClasses must exist (cluster-wide scope).
++ No node affinity or other requirements exist.
++ Operations must be idempotent in order to handle failure recovery.
+  
+## Workflows
+
+### Create Bucket
+1. The user creates an OBC in the app’s namespace.
+1. The COSI central controller detects the OBC and creates an OB (cluster-wide), containing driver-relevant data collected from the OBC and BC.
+1. The Sidecar detects the OB and issues a CreateBucket() rpc, passing to the driver config data.
+1. The Driver creates the bucket and returns pertinent endpoint, credential, and metadata.
+1. The Sidecar creates a secret in its namespace ("cosi-system") containing essential connection information.
+1. The central controller copies the secret to the OBC’s namespace and the app is ready to run.
```

Comment:
Please make sure to consider credentialing mechanisms that don't need explicit secrets.  E.g. when the existing kubernetes ServiceAccount is sufficient to access a bucket.

This will require information like "which Kube SA" and will not need information like "which secret"

Reply:
@thockin Thanks pointing this out - I hadn't realized that external services might use SAs as a means of authentication but it would simply workflow management.

Reply:
Google Cloud's Workload Identity operates like this.  I don't want to bias the design to my employer's needs, but I expect it to be a more common pattern over time.